File: app/src/main/cpp/wn-steam-client/src/job_manager.cpp
#include "wn_steam/job_manager.h"

#include <android/log.h>

#include <chrono>
#include <vector>

namespace wn_steam {

namespace {
constexpr const char* kLogTag = "WnSteamJM";
#define WN_LOGE(...) __android_log_print(ANDROID_LOG_ERROR, kLogTag, __VA_ARGS__)
#define WN_LOGD(...) __android_log_print(ANDROID_LOG_DEBUG, kLogTag, __VA_ARGS__)

// Seconds since UNIX epoch at process start. Used to encode an
// approximately-unique upper bits portion of the JobID so jobids
// across multiple client sessions don't collide if the server happens
// to remember any.
uint64_t make_process_epoch() {
    using namespace std::chrono;
    return static_cast<uint64_t>(
        duration_cast<seconds>(system_clock::now().time_since_epoch()).count());
}
}  // namespace

JobManager::JobManager(std::chrono::seconds default_timeout)
    : next_counter_(1),
      process_epoch_(make_process_epoch()),
      default_timeout_(default_timeout) {
    timeout_thread_ = std::thread(&JobManager::timeout_loop, this);
}

JobManager::~JobManager() {
    stop_.store(true);
    cv_.notify_all();
    if (timeout_thread_.joinable()) timeout_thread_.join();
    // Fail any remaining outstanding jobs so callers don't leak waits.
    fail_all("JobManager shutting down");
}

uint64_t JobManager::next_job_id() noexcept {
    const uint64_t lo = next_counter_.fetch_add(1, std::memory_order_relaxed);
    // Top 24 bits = process epoch, low 40 bits = counter. 40 bits gives
    // ~1 trillion jobids — practically unlimited for a session lifetime.
    return ((process_epoch_ & 0xFFFFFFull) << 40) | (lo & 0xFFFFFFFFFFull);
}

void JobManager::track(uint64_t job_id,
                       JobContinuation cb,
                       std::chrono::seconds timeout) {
    const auto t = (timeout.count() > 0) ? timeout : default_timeout_;
    Entry e{std::move(cb), Clock::now() + t};
    {
        std::lock_guard<std::mutex> lk(mu_);
        pending_[job_id] = std::move(e);
    }
    cv_.notify_all();
}

void JobManager::deliver(uint64_t job_id_target,
                         int32_t eresult,
                         std::string error_message,
                         std::span<const uint8_t> body) {
    JobContinuation cb;
    {
        std::lock_guard<std::mutex> lk(mu_);
        auto it = pending_.find(job_id_target);
        if (it == pending_.end()) {
            __android_log_print(ANDROID_LOG_INFO, kLogTag,
                "deliver: no pending job for jobid_target=0x%llx (eresult=%d, "
                "body=%zu bytes); known pending count=%zu",
                static_cast<unsigned long long>(job_id_target),
                eresult, body.size(), pending_.size());
            return;
        }
        cb = std::move(it->second.cb);
        pending_.erase(it);
    }

    // Steam OMITS the eresult field on successful unified-method responses
    // — the proto default in our deserializer (-1) means "not present on
    // the wire," and we treat that as OK. Real failures will carry a
    // positive EResult value (Fail=2, InvalidPassword=5, etc.).
    if (eresult == -1) eresult = 1;  // EResult.OK
    JobResult r;
    r.eresult       = eresult;
    r.error_message = std::move(error_message);
    r.body.assign(body.begin(), body.end());
    if (cb) {
        try {
            cb(std::move(r));
        } catch (const std::exception& e) {
            WN_LOGE("job continuation threw: %s", e.what());
        } catch (...) {
            WN_LOGE("job continuation threw unknown exception");
        }
    }
}

void JobManager::fail_all(const std::string& reason) {
    std::unordered_map<uint64_t, Entry> drained;
    {
        std::lock_guard<std::mutex> lk(mu_);
        drained.swap(pending_);
    }
    for (auto& [_, e] : drained) {
        if (!e.cb) continue;
        JobResult r;
        r.eresult           = -1;
        r.error_message     = reason;
        r.synthetic_failure = true;
        try { e.cb(std::move(r)); } catch (...) {}
    }
}

void JobManager::timeout_loop() {
    while (!stop_.load()) {
        std::vector<std::pair<uint64_t, JobContinuation>> expired;

        {
            std::unique_lock<std::mutex> lk(mu_);
            // Sleep until the earliest job deadline, or indefinitely when no
            // job is pending. A fixed 1s tick here woke the CPU ~86,400x/day
            // for nothing — an idle session must not wake at all. track()
            // notifies cv_ whenever a job is added, so a fresh job is picked
            // up immediately even out of the indefinite wait.
            if (pending_.empty()) {
                cv_.wait(lk, [this]() {
                    return stop_.load() || !pending_.empty();
                });
            } else {
                Clock::time_point earliest = Clock::time_point::max();
                for (const auto& kv : pending_) {
                    if (kv.second.deadline < earliest) earliest = kv.second.deadline;
                }
                cv_.wait_until(lk, earliest, [this]() { return stop_.load(); });
            }
            if (stop_.load()) return;

            const auto now = Clock::now();
            for (auto it = pending_.begin(); it != pending_.end(); ) {
                if (it->second.deadline <= now) {
                    expired.emplace_back(it->first, std::move(it->second.cb));
                    it = pending_.erase(it);
                } else {
                    ++it;
                }
            }
        }

        for (auto& [job_id, cb] : expired) {
            if (!cb) continue;
            JobResult r;
            r.eresult           = -1;
            r.error_message     = "job timeout";
            r.synthetic_failure = true;
            try {
                cb(std::move(r));
            } catch (const std::exception& e) {
                WN_LOGE("timeout continuation threw: %s", e.what());
            } catch (...) {}
        }
    }
}

}  // namespace wn_steam

File: app/src/main/cpp/wn-steam-client/rust/src/job_manager.rs
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::{Arc, Condvar, Mutex};
use std::thread::{self, JoinHandle};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
use std::{panic, panic::AssertUnwindSafe};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct JobResult {
    pub eresult: i32,
    pub error_message: String,
    pub body: Vec<u8>,
    pub synthetic_failure: bool,
}

impl Default for JobResult {
    fn default() -> Self {
        Self {
            eresult: 2,
            error_message: String::new(),
            body: Vec::new(),
            synthetic_failure: false,
        }
    }
}

type JobContinuation = Box<dyn FnOnce(JobResult) + Send + 'static>;

struct Entry {
    cb: JobContinuation,
    deadline: Instant,
}

#[derive(Default)]
struct Pending {
    jobs: HashMap<u64, Entry>,
}

pub struct JobManager {
    next_counter: AtomicU64,
    process_epoch: u64,
    default_timeout: Duration,
    pending: Arc<(Mutex<Pending>, Condvar)>,
    stop: Arc<AtomicBool>,
    timeout_thread: Option<JoinHandle<()>>,
}

impl JobManager {
    pub fn new(default_timeout: Duration) -> Self {
        let pending = Arc::new((Mutex::new(Pending::default()), Condvar::new()));
        let stop = Arc::new(AtomicBool::new(false));
        let thread_pending = Arc::clone(&pending);
        let thread_stop = Arc::clone(&stop);
        let timeout_thread = thread::spawn(move || timeout_loop(thread_pending, thread_stop));
        Self {
            next_counter: AtomicU64::new(1),
            process_epoch: make_process_epoch(),
            default_timeout,
            pending,
            stop,
            timeout_thread: Some(timeout_thread),
        }
    }

    pub fn next_job_id(&self) -> u64 {
        let lo = self.next_counter.fetch_add(1, Ordering::Relaxed);
        ((self.process_epoch & 0xFF_FFFF) << 40) | (lo & 0xFF_FFFF_FFFF)
    }

    pub fn track<F>(&self, job_id: u64, cb: F, timeout: Option<Duration>)
    where
        F: FnOnce(JobResult) + Send + 'static,
    {
        let timeout = timeout.unwrap_or(self.default_timeout);
        let (mu, cv) = &*self.pending;
        let mut pending = mu.lock().expect("job manager poisoned");
        pending.jobs.insert(
            job_id,
            Entry {
                cb: Box::new(cb),
                deadline: Instant::now() + timeout,
            },
        );
        cv.notify_all();
    }

    pub fn deliver(
        &self,
        job_id_target: u64,
        mut eresult: i32,
        error_message: String,
        body: &[u8],
    ) {
        let cb = {
            let (mu, _) = &*self.pending;
            let mut pending = mu.lock().expect("job manager poisoned");
            pending.jobs.remove(&job_id_target).map(|entry| entry.cb)
        };
        if let Some(cb) = cb {
            if eresult == -1 {
                eresult = 1;
            }
            invoke_continuation(
                cb,
                JobResult {
                    eresult,
                    error_message,
                    body: body.to_vec(),
                    synthetic_failure: false,
                },
            );
        }
    }

    pub fn fail_all(&self, reason: &str) {
        let drained = {
            let (mu, _) = &*self.pending;
            let mut pending = mu.lock().expect("job manager poisoned");
            std::mem::take(&mut pending.jobs)
        };
        for (_, entry) in drained {
            invoke_continuation(
                entry.cb,
                JobResult {
                    eresult: -1,
                    error_message: reason.to_string(),
                    body: Vec::new(),
                    synthetic_failure: true,
                },
            );
        }
    }
}

impl Default for JobManager {
    fn default() -> Self {
        Self::new(Duration::from_secs(30))
    }
}

impl Drop for JobManager {
    fn drop(&mut self) {
        self.stop.store(true, Ordering::Relaxed);
        self.pending.1.notify_all();
        if let Some(handle) = self.timeout_thread.take() {
            let _ = handle.join();
        }
        self.fail_all("JobManager shutting down");
    }
}

fn timeout_loop(pending: Arc<(Mutex<Pending>, Condvar)>, stop: Arc<AtomicBool>) {
    let (mu, cv) = &*pending;
    loop {
        let mut guard = mu.lock().expect("job manager poisoned");
        while guard.jobs.is_empty() && !stop.load(Ordering::Relaxed) {
            guard = cv.wait(guard).expect("job manager poisoned");
        }
        if stop.load(Ordering::Relaxed) {
            return;
        }
        let earliest = guard
            .jobs
            .values()
            .map(|entry| entry.deadline)
            .min()
            .unwrap_or_else(Instant::now);
        let now = Instant::now();
        if earliest > now {
            let (g, _) = cv
                .wait_timeout(guard, earliest - now)
                .expect("job manager poisoned");
            guard = g;
        }
        if stop.load(Ordering::Relaxed) {
            return;
        }
        let now = Instant::now();
        let expired_ids: Vec<u64> = guard
            .jobs
            .iter()
            .filter_map(|(id, entry)| (entry.deadline <= now).then_some(*id))
            .collect();
        let expired: Vec<Entry> = expired_ids
            .into_iter()
            .filter_map(|id| guard.jobs.remove(&id))
            .collect();
        drop(guard);
        for entry in expired {
            invoke_continuation(
                entry.cb,
                JobResult {
                    eresult: -1,
                    error_message: "job timeout".to_string(),
                    body: Vec::new(),
                    synthetic_failure: true,
                },
            );
        }
    }
}

fn invoke_continuation(cb: JobContinuation, result: JobResult) {
    let _ = panic::catch_unwind(AssertUnwindSafe(|| cb(result)));
}

fn make_process_epoch() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::mpsc;

    #[test]
    fn deliver_maps_missing_eresult_to_ok() {
        let jm = JobManager::new(Duration::from_secs(30));
        let id = jm.next_job_id();
        let (tx, rx) = mpsc::channel();
        jm.track(id, move |result| tx.send(result).unwrap(), None);
        jm.deliver(id, -1, String::new(), b"body");
        let result = rx.recv_timeout(Duration::from_secs(1)).unwrap();
        assert_eq!(result.eresult, 1);
        assert_eq!(result.body, b"body");
    }

    #[test]
    fn timeout_fails_jobs_synthetically() {
        let jm = JobManager::new(Duration::from_millis(20));
        let id = jm.next_job_id();
        let (tx, rx) = mpsc::channel();
        jm.track(id, move |result| tx.send(result).unwrap(), None);
        let result = rx.recv_timeout(Duration::from_secs(2)).unwrap();
        assert!(result.synthetic_failure);
        assert_eq!(result.error_message, "job timeout");
    }

    #[test]
    fn continuation_panic_is_caught() {
        let jm = JobManager::new(Duration::from_secs(30));
        let id = jm.next_job_id();
        jm.track(
            id,
            |_| {
                panic!("continuation panic");
            },
            None,
        );
        jm.deliver(id, 1, String::new(), b"body");
    }
}
